editor/components/inner-blocks/index.js

```diff
@@ -1,50 +1,110 @@
 /**
  * External dependencies
  */
+import { isEqual, pick } from 'lodash';
 import classnames from 'classnames';
 
 /**
  * WordPress dependencies
  */
 import { withContext } from '@wordpress/components';
 import { withViewportMatch } from '@wordpress/viewport';
-import { compose } from '@wordpress/element';
-import { withSelect } from '@wordpress/data';
+import { Component, compose } from '@wordpress/element';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { synchronizeBlocksWithTemplate } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
+import BlockList from '../block-list';
+import { withBlockEditContext } from '../block-edit/context';
 
-function InnerBlocks( {
-	BlockList,
-	layouts,
-	allowedBlocks,
-	template,
-	isSmallScreen,
-	isSelectedBlockInRoot,
-} ) {
-	const classes = classnames( 'editor-inner-blocks', {
-		'has-overlay': isSmallScreen && ! isSelectedBlockInRoot,
-	} );
-
-	return (
-		<div className={ classes }>
-			<BlockList { ...{ layouts, allowedBlocks, template } } />
-		</div>
-	);
+class InnerBlocks extends Component {
+	componentWillReceiveProps( nextProps ) {
+		this.updateNestedSettings( {
+			supportedBlocks: nextProps.allowedBlocks,
+		} );
+	}
+
+	componentDidMount() {
+		this.updateNestedSettings( {
+			supportedBlocks: this.props.allowedBlocks,
+		} );
+		this.insertTemplateBlocks( this.props.template );
+	}
+
+	insertTemplateBlocks( template ) {
+		const { block, insertBlocks } = this.props;
+		if ( template && ! block.innerBlocks.length ) {
+			// synchronizeBlocksWithTemplate( [], template ) parses the template structure,
+			// and returns/creates the necessary blocks to represent it.
+			insertBlocks( synchronizeBlocksWithTemplate( [], template ) );
+		}
+	}
+
+	updateNestedSettings( newSettings ) {
+		if ( ! isEqual( this.props.blockListSettings, newSettings ) ) {
+			this.props.updateNestedSettings( newSettings );
+		}
+	}
+
+	render() {
+		const {
+			uid,
+			renderBlockMenu,
+			layouts,
+			allowedBlocks,
+			template,
+			isSmallScreen,
+			isSelectedBlockInRoot,
+		} = this.props;
+
+		const classes = classnames( 'editor-inner-blocks', {
+			'has-overlay': isSmallScreen && ! isSelectedBlockInRoot,
+		} );
+
+		return (
+			<div className={ classes }>
+				<BlockList
+					rootUID={ uid }
+					renderBlockMenu={ renderBlockMenu }
+					{ ...{ layouts, allowedBlocks, template } }
+				/>
+			</div>
+		);
+	}
 }
 
 InnerBlocks = compose( [
-	withContext( 'BlockList' )(),
-	withContext( 'uid' )(),
+	withBlockEditContext( ( context ) => pick( context, [ 'uid', 'renderBlockMenu' ] ) ),
```

Comment:
In #7199 @youknowriad proposes how to get rid of `renderBlockMenu` from the context. Would it be possible to pass down `uid` to `InnerBlocks` instead to avoid exposing it through `BlockEditContext`?

Reply:
>In #7199 @youknowriad proposes how to get rid of `renderBlockMenu` from the context. Would it be possible to pass down `uid` to `InnerBlocks` instead to avoid exposing it through `BlockEditContext`?

Do you mean pass down as in by the developer in their own rendering of the component? Technically it's possible because we _do_ pass the UID as a prop to `edit`, but there was a bit of quality-of-life magic meant to be provided by the `InnerBlocks`. The component also lives inside `editor` so it doesn't seem like it should be unadvisable to rely on `withBlockEditContext` ?

Reply:
> The component also lives inside editor so it doesn't seem like it should be unadvisable to rely on withBlockEditContext ?

Right, I saw a file from `core-blocks` updated and assumed this one belongs there, too. My bad, please skip my comment :)